docs/foreign_code.md

```diff
@@ -0,0 +1,156 @@
+ - Most libraries are stateful, the correct granularity for interoperability is a library, not a function.
+
+The first of these relates to one of the core principles of Verona: there is no `unsafe` keyword.
+Verona targets security-critical workloads and it's very hard to make strong security claims about a language if any call to another language is able to compromise any and all of our safety guarantees.
+The Java JNI specification, for example, contains this disclaimer in the Design Overview chapter:
+
+> The programmer must not pass illegal pointers or arguments of the wrong type to JNI functions. Doing so could result in arbitrary consequences, including a corrupted system state or VM crash.
+
+The language makes an explicit declaration that any call to native code can undermine any of the invariants that the Java security model depends on. 
+This is a pragmatic decision to allow quick adoption but it means that no Java program can be considered safe until all native components have been removed.
+
+Modern C++ codebases work hard to provide memory safety guarantees at the API level.
+These guarantees are often lost in interop layers.
+For example, LLVM makes extensive use of typed smart pointers in the C++ APIs, but the Python bindings are written on top of the C API, which make type safety the caller's responsibility.
+Even if we can sandbox the resulting code, using a C++ library from Verona should not be more likely to introduce security vulnerabilities than using the same library from C++.
+The 2017 paper [*Finding and Preventing Bugs in JavaScript Bindings*](https://ieeexplore.ieee.org/document/7958598) showed that memory safety bugs in the C++ / JavaScript interop layer in Chrome are surprisingly common.
+Any time that a user has to hand-write code at a bindings layer to get memory safety right, we can assume that they will get it wrong and so we aim to automate as much as possible.
+
+The desire for sandboxing also leads to the conclusion that a 'foreign function interface' is exposing the wrong level of abstraction.
+What sandbox should a function run in?
+Our prior [CHERI-JNI](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf) work showed that treating libraries (in the loose sense, including any dependent libraries) as the unit of foreign code provided an abstraction that fitted well with sandboxing and allowed the user to determine the degree of state sharing between foreign code invocations.
+This abstraction also has software engineering benefits.
+LLVM, for example, does not have stable ABI (or API) guarantees across programs and so can cause linking errors if, for example, an application depends on one library that uses libclang from LLVM 10 to parse syntax highlighting and another that uses the JIT from LLVM 11 for GPU shaders.
+In a Verona world, these would be in different sandboxes with different 'global' namespaces and so would not conflict.
```

Comment:
I feel like this is touching the surface of something and then glides along.  I'm not sure what you re getting at, but it feels very important.

Reply:
Basically, there is no global namespace.  If you import a shared library, each instantiation of that library will see its own copy of all globals.  If you import two versions of the same library in different sandboxes, their namespaces are isolated as well as their heaps.  This prevents a bunch of annoying things that come from the C/UNIX dynamic linkage model.  Glibc and Android have some support in their linkers for loading libraries in different namespaces to address this - in prototyping, we should definitely use that, so even without sandboxing we don't touch the global namespace by accident.